File: python/bubble_sort.py
from ordering_enum import Ordering
import sys

def sort(array:list, order:Ordering=Ordering.ASC) -> list:
    '''
        Simplest sorting algorithm that works by repeatedly swapping the 
        adjacent elements if they are in wrong order. 
        Time Complex:
            - Best    -> O(n)
            - Average -> O(n^2)
            - Worst   -> O(n^2)
        Space Complex (Auxiliary Space): O(1)
        Stable: Yes
    '''
    # Exists the program if the ordering is not valid. 
    if (order not in [Ordering.ASC, Ordering.DESC]):
        sys.exit("Not Valid Ordering")

    # Traverse the array of elements
    num_elements:int = len(array)
    for i in range(num_elements):
        swapped:bool = False

        # The last element should already be in place
        for j in range(num_elements - i - 1):
            
            # Ordering preference selection
            if   (order == Ordering.ASC):  # ASCENDING
                if (array[j] > array[j + 1]):
                    # Swap elements if second element is greater than first 
                    # element
                    array[j], array[j + 1] = array[j + 1], array[j]
                    swapped = True
            elif (order == Ordering.DESC): # DESCENDING
                if (array[j] < array[j + 1]):
                    # Swap elements if second element is smaller than first 
                    # element
                    array[j], array[j + 1] = array[j + 1], array[j]
                    swapped = True

        # HINT: If nothing is swapped, elements are already ordered. Therefore, 
        # no further ordering operations are needed
        if not swapped:
            break

    return array        
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   

File: java/src/algorithms/BubbleSort.java
package src.algorithms;

import src.strategy.Order;

/**
 *
 * Simplest sorting algorithm, that works by repeatedly swapping the adjacent
 * elements, if they are in the wrong order.
 *
 * @author: <a href="mailto:p.aceredag@gmail.com">Pablo Acereda</a>
 * @version: 1.0
 * @license: Copyright 2021 © Pablo Acereda
 * License under Apache License, Version 2.0
 * 
 * @param <T>
 *
 */
public class BubbleSort<T> implements SortingAlgorithm<T> {
	/**
     * Sort an array of elements using Bubble sort algorithm.
     * 
     * @TimeComplexity  
     * Best    -> O(n) <br>
     * Average -> O(n^2) <br>
     * Worst   -> O(n^2) <br>
     * 
     * @SpaceComplexity
     * O(1)
     * 
     * @Stable
     * Yes
     *
     * @param array Elements to be ordered.
     * @param order Order preference.
     *
     * @return      Sorted array.
     */
	@Override
	public <T extends Comparable<T>> T[] sort(T[] elements, Order order) {		
		int numElements = elements.length;
		boolean swapped;
		// Iterate through all array
		for (int i=0; i < numElements - 1; i++) {
			swapped = false;
			// Iterate through the disordered part of the array (left-most part
			// should be ordered)
			for(int j=0; j < numElements-i-1; j++) {
				T current = elements[j];
				T next    = elements[j+1];
				
				if ((order == Order.ASC && current.compareTo(next) > 0)   || 
					(order == Order.DESC && current.compareTo(next) < 0)) {
					swap(elements, j, j+1);
					swapped = true;
				} 
			}
			
			if(!swapped) break;
		}
		
		return elements;
	}
	
	/**
	 * 
	 * @param <T> Generic type array.
	 * @param array Elements.
	 * @param i Position of first element to swap.
	 * @param j Position of second element to swap.
	 */
	private <T> void swap(T[] array, int i, int j) {
		T temporal = array[i];
		array[i] = array[j];
		array[j] = temporal;
	}

}